skcosmo/feature_selection/voronoi_fps.py

```diff
@@ -0,0 +1,136 @@
+
+        # self.voronoi_number should coincide with self.selected_idx_
+        # self.furthest_point = np.full(self.n_features_to_select, np.inf)
+
+        self.voronoi_location = np.full(n_features, 0)
+        self.idx_in_voronoi = [np.array([]) for n in range(self.n_features_to_select)]
+        self.idx_in_voronoi[0] = np.arange(n_features, dtype=int)
+        self.number_in_voronoi = np.full(self.n_features_to_select, 0)
+        self.number_in_voronoi[0] = n_features
+
+        # define the voronoi_r2 for the selected_ point
+        # this is the maximum distance from the center for each of the cells
+        # self.furthest_distance = np.zeros(self.n_features_to_select, float)
+
+        # quarter of the square distance between new selected point and previously
+        # selected points
+        self.sel_d2q_ = np.zeros(self.n_features_to_select, float)
+
+        super()._init_greedy_search(X, y, n_to_select)
+
+        # self.voronoi_r2[0] = self.haussdorf_[self.voronoi_i_far[0]]
+
+    def _get_active(self, X, last_selected):
+        # because we haven't called super() yet, self.n_selected_ has not been
+        # incremented yet
+
+        f_active = np.zeros(self.n_features_to_select, bool)
+
+        """must compute distance of the new point to all the previous FPS. Some
+            of these might have been computed already, but bookkeeping could be
+            worse that recomputing (TODO: verify!)
+            """
+        # calculation in a single block
+        self.sel_d2q_[: self.n_selected_] = (
+            self.norms_[self.selected_idx_[: self.n_selected_]]
+            + self.norms_[last_selected]
+            - 2 * (self.X_selected_[:, : self.n_selected_].T @ X[:, last_selected])
+        ) * 0.25
+
+        for ic in range(self.n_selected_):
+            # empty voronoi, no need to consider it
+            if self.number_in_voronoi[ic] > 1:
+                r2 = max(self.haussdorf_[self.idx_in_voronoi[ic]])
+                if self.sel_d2q_[ic] < r2:
+                    # these voronoi cells need to be updated
+                    f_active[ic] = True
+        f_active[self.n_selected_] = True
+        return f_active
+
+    def _move_point(self, point, new_voronoi_idx):
+        old_voronoi_loc = self.voronoi_location[point]
+        self.voronoi_location[point] = new_voronoi_idx
+
+        self.number_in_voronoi[old_voronoi_loc] -= 1
+        self.idx_in_voronoi[old_voronoi_loc] = np.array(
+            np.setdiff1d(self.idx_in_voronoi[old_voronoi_loc], [point]), dtype=int
+        )
+        self.number_in_voronoi[new_voronoi_idx] += 1
+
+        self.idx_in_voronoi[new_voronoi_idx] = np.array(
+            np.concatenate((self.idx_in_voronoi[new_voronoi_idx], [point])), dtype=int
+        )
+
+    def _repartition(self, X, old_voronoi_idx, new_voronoi_idx):
+
+        points_in_voronoi = self.idx_in_voronoi[old_voronoi_idx]
+        true_idx = self.selected_idx_[new_voronoi_idx]
+
+        new_dist = (
```

Comment:
In the previous iteration of code, we had additional check `if self.sel_d2q_[self.voronoi_number[j]] < self.haussdorf_[j]:` before making a calculation of `new_dist`. After this check we can drastically reduce calculating points.  Or you tried to do it and it didn't help?